### DISCUSSION TITLE
How to apply JSON patch locally?

### DISCUSSION THREAD

**Discussion by Yunabell-VU** [eclipse-emfcloud/emfcloud#184] [All other] [answered]
Hi, I'm currently working on a project that connects the local eclipse EObject model at the editor domain to the cloud server. I'm now able to send JSON patches to the server via PATCH call and receive JSON patches from the server through WebSockets. I wonder if there already exists a way from the EMF cloud server or example client implementation that I can use to apply the JSON patch to my local EObject, so the model can be synchronized with it on the server? 

I find out that there's a JsonPatchHelper class but I don't know if it can help with my implementation and if it does, how can I instantiate a JsonPatchHelper? Is there an example of using this helper?

Or, are there any other methods to apply JSON patches to EObject at my local editor domain?

Thank you!

**Comment by eneufeld** [comment, marked as answer]
Hi @Yunabell-VU ,

Yes the modelserver allows to modify the model using patches. You should use the Modelserver client, if you don't do so yet, to send patches to the modelserver.
You can find an example in the coffee editor for this. The relevant code for you is this method: https://github.com/eclipse-emfcloud/coffee-editor/blob/master/client/coffee-editor-extension/src/browser/coffee-tree/coffee-tree-editor-widget.tsx#L212

So the idea is, create a patch manually or using library like fast-json-patch. 
Then use the ModelServerClient `@eclipse-emfcloud/modelserver-client` to edit the model.

If you need more information let us know.

**Reply by Yunabell-VU** [reply]
Hi @eneufeld, 
Thanks for the reply. I'm still a bit confused about the mechanism. If I understand correctly, I can use the `edit` function provided by the `modelserver-client` to edit my local model directly, is that correct?

I checked the edit function from `ModelServerClientV2`,

```java
   public CompletableFuture<Response<String>> edit(final String modelUri, final ArrayNode jsonPatch,
      final String format) {
      String checkedFormat = delegate.checkedFormat(format);
      final Request request = new Request.Builder()
         .url(
            delegate.createHttpUrlBuilder(delegate.makeUrl(MODEL_BASE_PATH))
               .addQueryParameter(ModelServerPathParametersV2.MODEL_URI, modelUri)
               .addQueryParameter(ModelServerPathParametersV2.FORMAT, checkedFormat)
               .build())
         .patch(
            RequestBody.create(
               Json.object(
                  Json.prop(JsonRequestMember.DATA, Json.object(
                     Json.prop(JsonRequestMember.TYPE, Json.text(ModelServerPathParametersV2.JSON_PATCH)),
                     Json.prop(JsonRequestMember.DATA, Json.text(encode(jsonPatch))))))
                  .toString(),
               MediaType.parse("application/json")))
         .build();

      return makeCallAndGetDataBody(request);
   }
```
The thing I don't understand is, the function seems to build a request based on the JSON patch nodes, how does it modify my local model using the JSON patch received from the server? 
I'm sorry for the lame questions.

**Reply by eneufeld** [reply]
Hi @Yunabell-VU ,
Maybe I misunderstood you.

We usually treat the ModelServer as the source of truth, thus we usually override the local state when the modelserver sends us a new state.
So the edit function is used to notify the modelserver about local changes to update the remote state.
To listen to changes propagated by the modelserver there is a subscription service: 
`import { ModelServerSubscriptionServiceV2 } from '@eclipse-emfcloud/modelserver-theia/lib/browser';`
You can subscribe to it and you will get notified about changes.
Subscription:
https://github.com/eclipse-emfcloud/coffee-editor/blob/master/client/coffee-editor-extension/src/browser/coffee-tree/coffee-tree-editor-widget.tsx#L84
and the handling:
https://github.com/eclipse-emfcloud/coffee-editor/blob/master/client/coffee-editor-extension/src/browser/coffee-tree/coffee-tree-editor-widget.tsx#L120

Internally fast-json-patch#applyPatch is used if I remember correctly.

**Reply by Yunabell-VU** [reply]
Hi @eneufeld ,
Thanks for the explanation, now it's much clearer.
